martha/martha_v3.js

```diff
@@ -471,30 +289,23 @@ function validateRequest(url, auth, requestedFields) {
     }
 }
 
-/**
- * Used to check if any of the requested fields overlap with fields dependent on an underlying service.
- *
- * @param {string[]} requestedFields
- * @param {string[]} serviceFields
- * @returns {boolean} true if the requested fields overlap
- */
-function overlapFields(requestedFields, serviceFields) {
-    // via https://medium.com/@alvaro.saburido/set-theory-for-arrays-in-es6-eb2f20a61848
-    return requestedFields.filter((field) => serviceFields.includes(field)).length !== 0;
-}
-
 function buildRequestInfo(params) {
     const {
         url,
         requestedFields,
         auth,
+        forceAccessUrl,
     } = params;
 
     validateRequest(url, auth, requestedFields);
-    const drsType = determineDrsType(url);
+    const urlParts = getHttpsUrlParts(url);
+    const drsProvider = determineDrsProvider(url, urlParts);
+    // Force the retrieval of a (signed) access URL for this request if the `martha-force-access-url` header is set.
+    Object.assign(drsProvider, {forceAccessUrl: Boolean(forceAccessUrl)});
```

Comment:
Since `determineDrsProvider` reuses object instances, and since I'm pretty sure Cloud Function instances can service multiple requests, using `Object.assign` here could leak `forceAccessUrl` from one request to another.

Oh, but maybe a Cloud Function instance only handles one request at a time, so as long as `Boolean(forceAccessUrl)` evaluates to something then`forceAccessUrl` will have the correct value for _this_ invocation. It still makes me a little uncomfortable. I'd rather we deal in disposable instances of `DrsProvider`.

Also, I prefer `!!forceAccessUrl` (which I think of as "coerce to boolean") instead of `Boolean(forceAccessUrl)`. The double-negative felt like a little bit of a dirty hack to me at first, but I got used to it pretty quickly. `Boolean(...)`, which is correct, looks too much like `new Boolean(...)`, which is incorrect; from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean:
```
var x = Boolean(expression);     // use this...
var x = !!(expression);          // ...or this
var x = new Boolean(expression); // don't use this!
```

Reply:
Yes it is my understanding that Cloud Functions only handles one request at a time. If the framework is actually reentrant we may need to audit our casual use of global variables...

`eslint` is configured in this repo with the `no-implicit-coercion` setting as described [here](https://eslint.org/docs/rules/no-implicit-coercion). I did write this with `!!` originally and was admonished to change it to `Boolean`. 🙂

Reply:
Yeah I'll look at not reusing these instances.